pyqmc/twists.py

```diff
@@ -19,44 +18,27 @@ def get_frac_twist(cell, S, twist):
     return frac_twist
 
 
-def check_equivalent(kpts, ind, tol=1e-6):
-    """Given the index of a kpt in kpts, return the indices of its equivalent k-points in kpts"""
-    equiv_ind = []
-    for i in range(len(kpts)):
-        kdiffs = np.mod(kpts[i] - kpts[ind] + 0.5, 1.0) - 0.5
-        if np.linalg.norm(kdiffs) < tol:
-            equiv_ind.append(i)
-    return equiv_ind
-
-
-def get_qmc_kpts(cell, S, frac_twist):
-    """Given the cell, supercell transformation matrix and fractional twist, return the required k-points by a QMC calculation on that supercell"""
-    twist = get_twist(cell, S, frac_twist)
-    supercell = pyqmc.supercell.get_supercell(cell, S)
-    qmc_kpts = pyqmc.supercell.get_supercell_kpts(supercell) + twist
-    return qmc_kpts
-
-
-def get_k_indices(cell, mf_kpts, kpts, tol=1e-6):
-    """Given a list of kpts, return inds such that mf_kpts[inds] is a list of kpts equivalent to the input list"""
-    kdiffs = mf_kpts[np.newaxis] - kpts[:, np.newaxis]
-    frac_kdiffs = np.dot(kdiffs, cell.lattice_vectors().T) / (2 * np.pi)
-    kdiffs = np.mod(frac_kdiffs + 0.5, 1.0) - 0.5
-    return np.nonzero(np.linalg.norm(kdiffs, axis=-1) < tol)[1]
-
-
-def available_twists(cell, mf, S):
-    """Given the primitive cell and mf object from a DFT calculation and a supercell transformation matrix, return all valid twists (fractional) for a QMC calculation"""
-    ind = 0
-    unique_frac_kpts = np.array([get_frac_twist(cell, S, kpt) for kpt in mf.kpts])
-    while ind < len(unique_frac_kpts):
-        equiv_ind = check_equivalent(unique_frac_kpts, ind)
-        if len(equiv_ind) > 1:
-            unique_frac_kpts = np.delete(unique_frac_kpts, equiv_ind[1:], 0)
-        ind += 1
-    avail_twists = []
-    for frac_twist in unique_frac_kpts:
-        qmc_kpts = get_qmc_kpts(cell, S, frac_twist)
-        if len(get_k_indices(cell, mf.kpts, qmc_kpts)) == qmc_kpts.shape[0]:
-            avail_twists.append(frac_twist)
-    return np.array(avail_twists)
+def create_supercell_twists(supercell, mf, tol=12):
+    """
+
+    Args: 
+    supercell: a supercell object made from pyqmc.supercell.make_supercell
+    mf: a mf object with KRKS
+    tol: number of decimals to round the k-points. Default should be ok for most meshes
+
+    Returns: 
+    dictionary: 
+       twists: a list of ks's
+       counts: number of k-points in the primitive cell that correspond to the twist
+       kinds: indices of the k-points that correspond to the twist
+    """
+    kpts = mf.kpts
+    super_reciprocal_vectors = supercell.reciprocal_vectors()
+    super_kpts, wraparound = pyqmc.pbc.enforce_pbc(super_reciprocal_vectors,kpts)
+    twists, indices, counts = np.unique(np.round(super_kpts,tol), axis=0, return_counts=True, return_inverse=True)
```

Comment:
i think it will take some extra work to select a particular twist for a calculation. with the previous version, you chose the twist, like (0,0,0.5) and gave it directly to the Slater object. Now you first need to figure out which index to pass to Slater, so we might want to add a function for that. (or the function could take an effective twist supercell S and give the indices of the twists, similar to get_supercell_kpts)

Reply:
@willwheelera Doesn't `twists` serves the same purpose of `avail_twists` in the previous version?

Reply:
yes, it's the same